lib/perl/Genome/InstrumentData/AlignmentResult.pm

```diff
@@ -1628,10 +1643,144 @@ sub verify_alignment_data {
+    }
+}
+
+
+sub bam_header_path {
+    return File::Spec->join(shift->output_dir, 'all_sequences.bam.header');
+}
+
+
+sub flagstat_path {
+    return File::Spec->join(shift->output_dir, 'all_sequences.bam.flagstat');
+}
+
+
+sub get_merged_bam_to_revivify_per_lane_bam {
+    my $self = shift;
+    my $merged_result = $self->get_smallest_merged_alignment_result($self->get_unarchived_merged_alignment_results);
+
+    unless ($merged_result) {
+        $merged_result = $self->get_smallest_merged_alignment_result($self->get_merged_alignment_results);
+        $merged_result->_auto_unarchive;
+    }
+
+    my $merged_bam = $merged_result->merged_alignment_bam_path;
+    unless (-s $merged_bam) {
+        die $self->error_message("Merged bam (%s) does not exist for merged result id (%s)", $merged_bam, $merged_result->id);
+    }
+    return $merged_bam;
+}
+
+sub get_merged_alignment_results {
+    my $self = shift;
+    # Always load from the database, since other merged results may have committed since we updated the UR cache
+    my @results = Genome::InstrumentData::AlignmentResult::Merged->load(
+        'inputs.value_id' => $self->instrument_data_id,
+        test_name => $self->test_name,
+    );
+    my @filtered_results = $self->filter_non_database_objects(@results);
+    return $self->filter_non_matching_results(@filtered_results);
+}
+
+# This was refactored out to override in the test - mock objects break this logic
+sub filter_non_database_objects {
+    my ($self, @results) = @_;
+    my @db_results;
+    for my $result (@results) {
+        if (UR::Context->current->object_exists_in_underlying_context($result)) {
+            push @db_results, $result;
+        }
+    }
+    return @db_results;
+}
+
+# This uses a sort of 'backwards lookup' from merged alignment results going back to per-lane alignment results.
+# This is nice because it allows us to keep the logic in one place rather than mirroring it here.
+sub filter_non_matching_results {
+    my ($self, @merged_results) = @_;
+
+    my @matching_results;
+    for my $merged_result (@merged_results) {
+        my @individual_results = $merged_result->collect_individual_alignments;
+        if (@individual_results) {
+            push @matching_results, $merged_result if grep{refaddr($_) == refaddr($self)}@individual_results;
+        }
+    }
+
+    return @matching_results;
+}
+
+sub get_unarchived_merged_alignment_results {
     my $self = shift;
+    my @merged = $self->get_merged_alignment_results;
+    my @unarchived;
+    for my $merged (@merged) {
+        unless ( grep { $_->is_archived } $merged->disk_allocations ) {
+            push @unarchived, $merged;
+        }
+    }
+    return @unarchived;
+}
+
+sub get_smallest_merged_alignment_result {
+    my ($self, @results) = @_;
+    return unless @results;
+    my $smallest = $results[0];
+
+    for my $result (@results) {
+        my @current_id = $result->instrument_data;
```

Comment:
This has been done and pushed. Please let me know if you like the new way better or if it still needs work! Thank you.